torchx/lib/torchx/backend.ex

```diff
@@ -0,0 +1,268 @@
+  end
+
+  def iota(%T{shape: {n}, type: type} = out, 0) do
+    NIF.arange(0, n, 1, torch_type(type)) |> from_ref(out)
+  end
+
+  def iota(%T{shape: shape, type: type} = out, axis) do
+    NIF.iota(shape, axis, torch_type(type)) |> from_ref(out)
+  end
+
+  @impl true
+  def random_uniform(%T{type: {s, _} = type, shape: shape} = out, min, max) when s in [:u, :s] do
+    NIF.randint(min, max, shape, torch_type(type)) |> from_ref(out)
+  end
+
+  def random_uniform(%T{type: {f, _} = type, shape: shape} = out, min, max) when f in [:f, :bf] do
+    NIF.rand(min, max, shape, torch_type(type)) |> from_ref(out)
+  end
+
+  @impl true
+  def random_normal(%T{type: _type, shape: shape} = out, mu, sigma) do
+    NIF.normal(mu, sigma, shape) |> from_ref(out)
+  end
+
+  ## Transfer
+
+  @impl true
+  def to_batched_list(%T{shape: shape} = out, %T{data: %TB{ref: ref}}),
+    do: NIF.split(ref, elem(shape, 0)) |> from_list_ref(out)
+
+  @impl true
+  def to_binary(_tensor, _limit \\ nil) do
+    raise "operation to_binary is not supported on Torchx.Backend. " <>
+            "You must first transfer the tensor to Elixir by calling Nx.backend_transfer/1"
+  end
+
+  defp to_blob(tensor, limit \\ nil)
+  defp to_blob(%T{data: %TB{ref: ref}}, nil), do: NIF.to_blob(ref)
+  defp to_blob(%T{data: %TB{ref: ref}}, limit), do: NIF.to_blob(ref, limit)
+
+  @impl true
+  def backend_deallocate(%Nx.Tensor{data: %TB{ref: ref}}), do: NIF.delete_tensor(ref)
+
+  @impl true
+  def backend_transfer(tensor, Nx.Tensor, opts) do
+    backend_transfer(tensor, Nx.BinaryBackend, opts)
+  end
+
+  def backend_transfer(tensor, Torchx.Backend, _opts) do
+    tensor
+  end
+
+  def backend_transfer(tensor, backend, opts) do
+    backend.from_binary(tensor, to_blob(tensor), opts)
+  end
+
+  @impl true
+  def from_binary(%T{type: type, shape: shape} = out, binary, _opts) do
+    NIF.from_blob(binary, shape, torch_type(type)) |> from_ref(out)
+  end
+
+  ## Shape
+
+  @impl true
+  def reshape(out, %T{data: %TB{ref: ref}}, shape),
+    do: NIF.reshape(ref, shape) |> from_ref(out)
+
+  @impl true
+  def as_type(%T{type: type} = out, %T{data: %TB{ref: ref}}),
+    do: NIF.to_type(ref, torch_type(type)) |> from_ref(out)
+
+  @impl true
+  def squeeze(out, tensor, _axes) do
+    NIF.squeeze(tensor.data.ref) |> from_ref(out)
+  end
+
+  @impl true
+  def transpose(out, tensor, axes), do: IO.inspect(axes)
+  # do: NIF.transpose(tensor, dim0, dim1) |> from_ref(out)
+
+  ## Ops
+
+  binary_ops =
+    [:add, :subtract, :multiply, :power, :remainder, :divide, :atan2, :min, :max, :quotient] ++
+      [:bitwise_and, :bitwise_or, :bitwise_xor, :left_shift, :right_shift] ++
+      [:equal, :not_equal, :greater, :less, :greater_equal, :less_equal] ++
+      [:logical_and, :logical_or, :logical_xor] ++
+      [:outer]
+
+
+  for op <- binary_ops do
+    @impl true
+    def unquote(op)(out, left, right) when is_number(right) do
```

Comment:
I don't think it will ever be a number. So we can likely remove this clause.

Reply:
So, all scalars in ops will be wrapped in tensors, like this?
```
iex(1)> Nx.tensor(5.0)
#Nx.Tensor<
  f32
  5.0
>
```

Reply:
Yes, I see I can:)
Me better do not write comments while still sleeping:)